### PR TITLE
v0.9

### PR DESCRIPTION
**Pull request by aegatlin**

### CODE REVIEW

**Comment by aegatlin** [comment]
@esakal I bumped the version to v0.8 for you changeset, plus some additional refactors on top. 

Here's a heads up on my current ideas for v0.9. I'm curious what you think.

This is a (poorly named) branch for v0.9. I want to lean more heavily into the general-purpose nature of metamark as a markdown manipulation utility while also making it easier to work with obsidian. The way I'm attempting to do this is by pulling things apart into namespaces. For example, assuming `m` was the metamark namespace in general, my idea is as follows:

```ts
const { filePaths, pageAllowSet } = m.obsidian.processVault("../vault");
const md = m.file.getMd.fromFilePath("../vault/file.md");
const md = m.file.getMd.fromContent("# Hello\n world");
const marks = m.mark.getMarks(...);
const firstPara = m.unified.getFirstParagraphText(md);
```

`m.obsidian` could encapsulate what is currently being done by consumers, simplifying the adoption workflow required.

All I'm really doing is moving the current functionality around into namespaces.

In order to facilitate this I'm upgrading the testing framework to vitest so I can unit test more easily. And moving to tsup, so I can bundle without worrying about exposing my dir structure. My goal is to mostly leave things unchanged, plus add novel functionality. If functionality should change as I work on this, I will try to leave that for v0.9.1.

**Comment by aegatlin** [comment]
I think it would also be a good idea for the mark-related utilities to receive a configuration object, that way there's one place for defining how one would customize, e.g., the link-creation logic. And it can also have the "reasonable defaults" to fallback on

**Comment by aegatlin** [comment]
@esakal How are you using this lib currently? I'm seeing in your tests that you have `frontmatter.slugBase` and then in the test suite you also have a `getSlug` function adding a custom suffix, resulting in a uri like the following: `/custom/base/wiki-link-custom-suffix`. I'm curious what you are _really_ up to with these in your app? 

1. Would a configuration options for a custom prefix work? E.g., the default points to `/content/...`, but a config option could be included to override that to `/custom/base/...`. Would that satisfy your use case?
2. Do you really use a custom suffix? What is the use case you want supported with a suffix?

Basically, I want to continue supporting your use cases in the refactor. But this refactor introduces major API changes, so I want to make sure I preserve the right use cases in the overhaul.

THank you 🙏

**Comment by aegatlin** [comment]
Okay, i have tried my best to preserve the behavior of the previous version, but this was a massive refactor with a focus on introducing the `obsidian.vault.process` function, which brings in house the previously manual step of leveraging the file parsing of older `metamark`. I think this is a good step because parts of the processing pipeline presumed data structures that only existed in obsidian (or other wiki-like tools).

In the future I think metamark can reintroduce per-file processing. In particular, such a one-off file could leverage a unified processing pipeline that does _not_ involve `remark-obsidian-link`, since standalone markdown files presumably don't reference wiki links.

One feature was left out from your changes (I believe). And that was the default `target=_blank` for anchor tags, which open links in new tabs. I was considering introducing a configuration option that was something like `shouldOpenLinksInNewTab: boolean` that triggered it. But, I was "bike shedding" over it and decided to drop the issue. The reasoning was that you can override the unified processor manually via `opts.unifiedProcessorBuilder`. But, that is a _large_ change, so it might not be ideal. I think the configuration can evolve going forward. If you have a strong opinion on it, let me know :D
